### PR TITLE
Checkpoints Simplified

### PR DESCRIPTION
**Pull request by mvpatel2000**
# What does this PR do?

Refactors checkpoint workflow to be simpler and cleaner. Fixes bug resulting in timeout with checkpoints.
Also, fixes bug with autoresume + sharded checkpointing.
Top lines are old way, bottom lines are corrected code:
image

Working with autoresume
image

# What issue(s) does this change relate to?

[CO-1881](https://mosaicml.atlassian.net/browse/CO-1881)
[CO-1883](https://mosaicml.atlassian.net/browse/CO-1883)

[CO-1881]: https://mosaicml.atlassian.net/browse/CO-1881?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[CO-1883]: https://mosaicml.atlassian.net/browse/CO-1883?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: composer/trainer/trainer.py [hunk 1]
```diff
@@ -1456,15 +1456,16 @@ def _get_autoresume_checkpoint(
             f'Looking for autoresume checkpoint: {save_latest_remote_file_name} (remote), {latest_checkpoint_path} (local)'
         )
 
-        if self.deepspeed_enabled:
+        # DeepSpeed and FSDP sharded state dict save checkpoints on every rank
+        if self.deepspeed_enabled or self.state.fsdp_sharded_state_dict_enabled:
```
Comment (dakinggg) [inline composer/trainer/trainer.py:1460]: did you test autoresume + FSDP sharded? I think that should be unit testable...

**Review by dakinggg — COMMENTED** [review]
Looks alright to me, leaving approval for @eracah

File: composer/trainer/trainer.py (hunk 1 above)
Comment (mvpatel2000) [reply composer/trainer/trainer.py:1460]: @eracah do u wanna add it :) I did not test it

Comment (eracah) [reply composer/trainer/trainer.py:1460]: ok added unit test for sharded + autoresume and fixed a bug the unit test discovered 😉

**Review by eracah — APPROVED** [review]
LGTM

**Pull request merged**